### ISSUE TITLE
Losing messages in RetryingBatchErrorHandler/ErrorHandlingUtils during rebalancing

### ISSUE THREAD

**Issue by vooft** [spring-projects/spring-kafka#2340]
**In what version(s) of Spring for Apache Kafka are you seeing this issue?**

For example:

2.3.7+

**Describe the bug**

If a listener doing a retry and rabalancing kicks in, then the consumer will be unpaused and every consumer.poll() invocation will start returning records and the error handler will throw them away. 

It seems that the rebalancing listener KafkaMessageListenerContainer can re-pause the consumer again, but it is not aware that the error handler paused it.

**To Reproduce**

Create a listener with infinite retry RetryingBatchErrorHandler that will be constantly failing and force rebalancing. After rebalancing finished RetryingBatchErrorHandler will drain all the messages from assigned partitions without any processing.

**Expected behavior**

RetryingBatchErrorHandler should re-pause consumer after rebalancing.

**Sample**

Test is using RetryingBatchErrorHandler, but logic in ErrorHandlingUtils is essentially the same
https://github.com/vooft/kafka-retry-issue/blob/master/src/test/java/com/example/kafkaissue/KafkaIssueTest.java

**Comment by vooft** [comment]
Thank you for your answer @garyrussell 
I'm looking at `DefaultErrorHander` and it seems that if you just throw an arbitrary exception from the listener it will still go all the way to `ErrorHandlingUtils` and will have the same issue as in `RetryingBatchErrorHandler`, or am I missing something?

**Comment by garyrussell** [comment]
Yes; I agree; I was just pointing out that the recovering mechanism is preferred to the (older) retrying mechanism that suffers from this bug. I will have a fix soon.

**Comment by garyrussell** [comment]
@vooft 2.8.8 with the fix is now in Maven Central.
